src/tir/transforms/storage_flatten.cc

```diff
@@ -1160,18 +1144,70 @@ class BufferBindUnwrapper : public StmtExprMutator {
+    if (buf_map_.count(buffer.get())) {
+      const BufferEntry& e = buf_map_[buffer.get()];
+      ICHECK(e.in_scope) << "Cannot access a buffer " << buffer->name << ", out of scope";
+      return e;
+    } else if (buffer_var_defines_.count(buffer->data.get())) {
+      // The buffer var was defined, but the buffer hasn't been seen
+      // before.
       BufferEntry entry;
       entry.buffer = buffer;
+      var_to_buffer_[buffer->data.get()] = buffer;
       buf_map_[buffer.get()] = std::move(entry);
-    }
+      return buf_map_[buffer.get()];
+    } else if (var_remap_.count(buffer->data.get())) {
+      // The buffer var is an alias of a bound buffer.  Only
+      // supported if the bound buffer has no offsets.  In this
+      // case, we just need to make a new aliasing buffer that
+      // shares the remapped data variable.
+      Var old_var = buffer->data;
+      Var new_var = Downcast<Var>(var_remap_[old_var.get()]);
 
-    auto it = buf_map_.find(buffer.get());
-    ICHECK(it != buf_map_.end()) << "Cannot find allocated buffer for " << buffer;
-    const BufferEntry& e = it->second;
-    ICHECK(e.in_scope) << "Cannot access a buffer " << buffer->name << ", out of scope";
-    return it->second;
+      {
+        ICHECK(var_to_buffer_.count(old_var.get()))
+            << "Cannot find remap information for aliased buffer var " << old_var->name_hint
+            << ", required to verify this alias is legal.";
+        const Buffer& aliased_buffer = var_to_buffer_[old_var.get()];
+        const BufferEntry& entry = buf_map_[aliased_buffer.get()];
+        if (entry.remap) {
+          for (const auto& begin : entry.remap->begins) {
+            ICHECK(is_zero(begin)) << "Aliasing of buffer with offset is not supported";
+          }
+        }
+      }
+
+      {
+        Buffer new_buf = buffer;
+        new_buf.CopyOnWrite()->data = new_var;
+
+        RemapInfo remap_info;
+        remap_info.target = new_buf;
+        remap_info.begins = Array<PrimExpr>(buffer->shape.size(), 0);
+        remap_info.extents = buffer->shape;
+
+        BufferEntry entry;
+        entry.buffer = buffer;
+        entry.remap = std::make_unique<RemapInfo>(remap_info);
+        entry.in_scope = true;
+        var_to_buffer_[buffer->data.get()] = buffer;
+        buf_map_[buffer.get()] = std::move(entry);
+      }
+      return buf_map_[buffer.get()];
+    } else if (var_to_buffer_.count(buffer->data.get())) {
+      // This buffer is an alias of a known buffer, with no remaps.  A
+      // buffer entry should be generated and returned.
+      BufferEntry entry;
+      entry.buffer = buffer;
+      entry.in_scope = true;
+      var_to_buffer_[buffer->data.get()] = buffer;
+      buf_map_[buffer.get()] = std::move(entry);
+
+      return buf_map_[buffer.get()];
+    } else {
+      LOG(FATAL) << "Can't work around the undefined buffer";
+      return *static_cast<BufferEntry*>(nullptr);
```

Comment:
This is not good, please use `__builtin_unreachable()` or something like that...

Reply:
Given that the effect of control flow reaching `__builtin_unreachable()` is to make the program behavior undefined, I had assumed that intentionally invoking undefined behavior in an unreachable location would be identical to  `__builtin_unreachable()`, without relying on non-standard intrinsics.  The C++23 implementation of `std::unreachable()` has the same semantics, that program behavior is undefined if the unreachable location is reached.

That said, this pattern is primarily to suppress the warning that occurs if a return value isn't in a branch, and isn't the most readable.  Testing, it looks like a much better way to suppress the warning would be to mark `LOG(FATAL)` as `[[noreturn]]`, so that the warning never needs to be suppressed in the first place.

Reply:
Looks like there were two independent `LOG(FATAL)` implementations, and only the one that occurred when `TVM_LOG_CUSTOMIZE` was defined has `[[noreturn]]` applied to it.

Reply:
> No matter what, as it is here, this is an explicit UB.

Agreed.  That was intentional as a way to indicate an unreachable path without compiler-specific intrinsics, though I'm realizing that my intent for it to be both (1) deducible by the compiler as undefined for the purpose of optimization and (2) not deducible by the compiler as undefined for the purpose of producing warnings wasn't the most consistent of plans.

Reply:
I've opened https://github.com/apache/tvm/pull/13542, which removes this line entirely, with the `[[noreturn]]` attribute applied to both implementations of `LOG(FATAL)` so that the warning does not occur.